portal-history: separate styles

johnd0e opened the pull request:
Separate style for every included here highlighter.
Any style may be overridden without touching main logic:
- inline (`portalsHistory.styles` object)
- or in separate plugin, with code like this:
  ```js
  var setup = function () {
    var portalsHistory = window.plugin.portalHighlighterPortalsHistory;
    if (!portalsHistory || !portalsHistory.styles) {
      console.warn('highlight-portal-history plugin is required!');
      return;
    }
    portalsHistory.styles.common.radius = 10;
    portalsHistory.styles.commonOther.fillOpacity = 0.1;
  };
  ```

Known styles:
```
  captured <------------ |
  visitTarget <--------- |
  scoutControlled <----- |
  scoutControllTarget <- | <== marked ===== |
                                            | <= common
  visited <------------- | <== semiMarked = |
  captureTarget <------- |
```
```
  otherVC <------- | <= commonOther (empty by default)
  otherScout <---- |
  otherNotVC <---- |
  otherNotScout <- |
```

Commit:
highlight-portal-history: minor refactoring

Commit:
highlight-portal-history: separate overridable styles

MysticJay commented:
The code is complicated without need.
At two places a new function gets inserted that only call one global function.
The construction of the styles from various parts is complex and not easy to maintain.

Create direct named style-objects that hold all wanted properties
and use them directly in `data. portal.setStyle(<name of style>)`

johnd0e commented:
`notVisited` has a bug now: we need to check `!visited` first, then `!captured`.
Will fix tomorrow.

Commit:
highlight-portal-history: implement style inheritance

  captured <------------ |
  visitTarget <--------- |
  scoutControlled <----- |
  scoutControllTarget <- | <== marked ===== |
                                            | <= common
  visited <------------- | <== semiMarked = |
  captureTarget <------- |

Sample plugin code to change radius of every style:
```
var setup = function () {
  var portalsHistory = window.plugin.portalHighlighterPortalsHistory;
  if (!portalsHistory || !portalsHistory.styles) {
    console.warn('highlight-portal-history plugin is required!');
    return;
  }
  portalsHistory.styles.common.radius = 10;
};
```

Commit:
highlight-portal-history: refactor style inheritance

Commit:
highlight-portal-history: optional styling of not matched portals [experiment]

  otherVC <------- | <= commonOther (empty by default)
  otherScout <---- |
  otherNotVC <---- |
  otherNotScout <- |

Test:
  window.plugin.portalHighlighterPortalsHistory.styles.commonOther.radius = 3;

modos189 commented on plugins/highlight-portal-history.js:
```diff
@@ -1,73 +1,104 @@
-  window.addPortalHighlighter('History: not visited/captured', window.plugin.portalHighlighterPortalsHistory.notVisited);
-  window.addPortalHighlighter('History: scout controlled', window.plugin.portalHighlighterPortalsHistory.scoutControlled);
-  window.addPortalHighlighter('History: not scout controlled', window.plugin.portalHighlighterPortalsHistory.notScoutControlled);
+function inherit (parentName, childNames) {
```
```suggestion
// Creating styles based on a given template
function inherit (parentName, childNames) {
```

modos189 commented on plugins/highlight-portal-history.js:
```diff
@@ -1,73 +1,104 @@
+function inherit (parentName, childNames) {
+  var styles = portalsHistory.styles;
+  childNames.forEach(function (name) {
+    styles[name] = L.extend(L.Util.create(styles[parentName]), styles[name]);
```
```suggestion
    // Extension of _styles_ with a new _name_ object, created based on _parentName_ object.
    styles[name] = L.extend(L.Util.create(styles[parentName]), styles[name]);
```
